datasources/src/main/scala/com/normation/plugins/datasources/QueryService.scala

```diff
@@ -83,6 +85,54 @@ trait QueryDataSourceService {
    * A version that only query one node - do not use if you want to query several nodes
    */
   def queryOne(datasource: DataSource, nodeId: NodeId, cause: UpdateCause): IOResult[NodeUpdateResult]
+
+  /*
+   * We can delete a datasource on all nodes
+   */
+  def deleteAll(datasource: DataSourceId, cause: UpdateCause): IOResult[Set[NodeUpdateResult]]
+}
+
+object QueryService {
+
+  /*
+   * Utility method that handle updating a node with given property value
+   * for given datasource
+   */
+  def updateNode(
+      nodeInfo:          NodeInfo,
+      repository:        WoNodeRepository,
+      optProperty:       Option[NodeProperty],
+      cause:             UpdateCause,
+      enforceSameOrigin: Boolean // for deletion, we want to be able to remove only prop set by datasource provider
+  ): IOResult[NodeUpdateResult] = {
+    optProperty match {
+      // on none, don't update anything, the life is wonderful (because 'none' means 'don't update')
+      case None           => NodeUpdateResult.Unchanged(nodeInfo.id).succeed
+      case Some(property) =>
+        // look for the property value in the node to know if an update is needed.
+        // we only care about value here (not provider or other meta-info)
```

Comment:
I believe this comment is no longer up-to-date with the new behavior : it looks like we also care about the provider when `enforceSameOrigin` is true

Reply:
yes